File: servidor.py
# servidor.py (adaptado)
import grpc
from concurrent import futures
import notas_pb2
import notas_pb2_grpc

# Armazenamento em memória (simulando um banco de dados)
db_notas = {} # Ex: { "123_DISC1": Nota(...), "123_DISC2": Nota(...) }

class GerenciadorNotasServicer(notas_pb2_grpc.GerenciadorNotasServicer):

    def AdicionarNota(self, request, context):
        print(f"Adicionando nota para RA {request.ra} na disciplina {request.cod_disciplina}")
        chave = f"{request.ra}_{request.cod_disciplina}"

        if chave in db_notas:
            return notas_pb2.StatusResponse(sucesso=False, msg="Nota já existe para este aluno/disciplina. Use AlterarNota.")

        # Criando a Nota
        nota = notas_pb2.Nota(
        ra=request.ra,
        cod_disciplina=request.cod_disciplina,
        ano=request.ano,
        semestre=request.semestre,
        nota=request.nota
        )
        db_notas[chave] = nota
        return notas_pb2.StatusResponse(sucesso=True, msg="Nota adicionada com sucesso!")

    def AlterarNota(self, request, context):
        print(f"Alterando nota para RA {request.ra} na disciplina {request.cod_disciplina}")
        chave = f"{request.ra}_{request.cod_disciplina}"

        if chave not in db_notas:
            return notas_pb2.StatusResponse(sucesso=False, msg="Nota não encontrada. Adicione a nota antes de alterar.")

# Alterando a nota
        nota = db_notas[chave]
        nota.nota = request.nota # Atualizando o valor da nota
        nota.ano = request.ano
        nota.semestre = request.semestre
        db_notas[chave] = nota
        return notas_pb2.StatusResponse(sucesso=True, msg="Nota alterada com sucesso!")

    def ConsultarNota(self, request, context):
        print(f"Consultando nota para RA {request.ra} na disciplina {request.cod_disciplina}")
        chave = f"{request.ra}_{request.cod_disciplina}"

        if chave not in db_notas:
            return notas_pb2.ConsultaNotaResponse(sucesso=False, msg_erro="Nota não encontrada.")

        # Retornando a nota consultada
        nota = db_notas[chave]
        return notas_pb2.ConsultaNotaResponse(sucesso=True, nota=nota, msg_erro="")

    def CalcularMedia(self, request, context):
        print(f"Calculando a média para RA {request.ra}")
        soma_notas = 0
        contador = 0

        # Itera pelas notas do aluno
        for chave, nota in db_notas.items():
            if nota.ra == request.ra:
                soma_notas += nota.nota
                contador += 1

        if contador == 0:
            return notas_pb2.MediaResponse(sucesso=False, media=0, msg_erro="Nenhuma nota encontrada para o RA informado.")

        media = soma_notas / contador
        return notas_pb2.MediaResponse(sucesso=True, media=media, msg_erro="")

    def ListarNotasAluno(self, request, context):
        """
        Este método utiliza streaming do lado do servidor.
        Ele envia as notas de um aluno uma por uma.
        """
        
        print(f"Listando notas para o RA {request.ra}")

        # Itera sobre o banco de dados e retorna as notas do aluno
        for chave, nota in db_notas.items():
            if nota.ra == request.ra:
                yield nota

            # Se o aluno não tiver notas, envia um erro
            if not any(nota.ra == request.ra for chave, nota in db_notas.items()):
                context.set_code(grpc.StatusCode.NOT_FOUND)
                context.set_details(f"Nenhuma nota encontrada para o RA {request.ra}")
                yield notas_pb2.Nota() # Retorna uma nota vazia como resposta


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    notas_pb2_grpc.add_GerenciadorNotasServicer_to_server(GerenciadorNotasServicer(), server)
    server.add_insecure_port('[::]:50052')
    server.start()
    print("Servidor gRPC rodando na porta 50052.")
    server.wait_for_termination()

if __name__ == '__main__':
    serve()

File: servidor.cpp
#include <iostream>
#include <memory>
#include <string>
#include <map>
#include <vector>

#include <grpcpp/grpcpp.h>
#include "notas.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;
using grpc::ServerWriter;

// Implementação do serviço
class GerenciadorNotasImpl final : public gerencia_notas::GerenciadorNotas::Service {
private:
    // Estrutura para armazenar notas em memória
    std::map<std::string, std::vector<gerencia_notas::Nota>> notas_por_aluno;
    
public:
    Status AdicionarNota(ServerContext* context, const gerencia_notas::AdicionaNotaRequest* request, gerencia_notas::StatusResponse* response) override {
        std::cout << "Adicionando nota para RA: " << request->ra() << std::endl;
        
        // Cria uma nova nota
        gerencia_notas::Nota nova_nota;
        nova_nota.set_ra(request->ra());
        nova_nota.set_cod_disciplina(request->cod_disciplina());
        nova_nota.set_ano(request->ano());
        nova_nota.set_semestre(request->semestre());
        nova_nota.set_nota(request->nota());
        
        // Adiciona à estrutura de dados
        notas_por_aluno[request->ra()].push_back(nova_nota);
        
        response->set_sucesso(true);
        response->set_msg("Nota adicionada com sucesso");
        
        return Status::OK;
    }

    Status AlterarNota(ServerContext* context, const gerencia_notas::AdicionaNotaRequest* request, gerencia_notas::StatusResponse* response) override {
        std::cout << "Alterando nota para RA: " << request->ra() << std::endl;
        
        // Busca a nota existente
        auto& notas = notas_por_aluno[request->ra()];
        bool encontrou = false;
        
        for (auto& nota : notas) {
            if (nota.cod_disciplina() == request->cod_disciplina() && 
                nota.ano() == request->ano() && 
                nota.semestre() == request->semestre()) {
                nota.set_nota(request->nota());
                encontrou = true;
                break;
            }
        }
        
        if (encontrou) {
            response->set_sucesso(true);
            response->set_msg("Nota alterada com sucesso");
        } else {
            response->set_sucesso(false);
            response->set_msg("Nota não encontrada");
        }
        
        return Status::OK;
    }

    Status ConsultarNota(ServerContext* context, const gerencia_notas::AlunoDisciplinaRequest* request, gerencia_notas::ConsultaNotaResponse* response) override {
        std::cout << "Consultando nota para RA: " << request->ra() << ", Disciplina: " << request->cod_disciplina() << std::endl;
        
        auto it = notas_por_aluno.find(request->ra());
        if (it != notas_por_aluno.end()) {
            for (const auto& nota : it->second) {
                if (nota.cod_disciplina() == request->cod_disciplina()) {
                    response->set_sucesso(true);
                    *response->mutable_nota() = nota;
                    return Status::OK;
                }
            }
        }
        
        response->set_sucesso(false);
        response->set_msg_erro("Nota não encontrada");
        return Status::OK;
    }

    Status CalcularMedia(ServerContext* context, const gerencia_notas::AlunoRequest* request, gerencia_notas::MediaResponse* response) override {
        std::cout << "Calculando média para RA: " << request->ra() << std::endl;
        
        auto it = notas_por_aluno.find(request->ra());
        if (it != notas_por_aluno.end() && !it->second.empty()) {
            float soma = 0.0;
            int count = 0;
            
            for (const auto& nota : it->second) {
                soma += nota.nota();
                count++;
            }
            
            float media = soma / count;
            response->set_sucesso(true);
            response->set_media(media);
        } else {
            response->set_sucesso(false);
            response->set_media(0.0);
            response->set_msg_erro("Nenhuma nota encontrada para o aluno");
        }
        
        return Status::OK;
    }

    Status ListarNotasAluno(ServerContext* context, const gerencia_notas::AlunoRequest* request, ServerWriter<gerencia_notas::Nota>* writer) override {
        std::cout << "Listando notas para RA: " << request->ra() << std::endl;
        
        auto it = notas_por_aluno.find(request->ra());
        if (it != notas_por_aluno.end()) {
            for (const auto& nota : it->second) {
                writer->Write(nota);
            }
        }
        
        return Status::OK;
    }
};

void RunServer() {
    std::string server_address("0.0.0.0:50052");
    GerenciadorNotasImpl service;

    ServerBuilder builder;
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);
    
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "Servidor rodando em " << server_address << std::endl;

    server->Wait();
}

int main(int argc, char** argv) {
    RunServer();
    return 0;
}